### PR TITLE
X axis zoom

### PR DESCRIPTION
**Pull request by stuartnelson3**
requested in #241

### CODE REVIEW

File: app/assets/javascripts/angular/directives/graph_chart.js [hunk 1]
```diff
@@ -11,6 +12,145 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
       var graphData = [];
       var annotationData = [];
 
+      // figure out how to move this into a separate service, and
```
Comment (grobie) [inline app/assets/javascripts/angular/directives/graph_chart.js]: Mark a todo with `// TODO(stuart)`.

File: app/assets/javascripts/angular/controllers/graph_ctrl.js [hunk 2]
```diff
@@ -123,7 +123,8 @@ angular.module("Prometheus.controllers").controller('GraphCtrl',
       var rangeSeconds = Prometheus.Graph.parseDuration(scope.graph.range);
       // bigger denominator == smaller step == more data
       var step = Math.floor(rangeSeconds / scalingFactor)
-      refreshFn(rangeSeconds, step).then(function(data) {
+      var endTime = Math.floor($scope.graph.endTime / 1000);
```
Comment (grobie) [inline app/assets/javascripts/angular/controllers/graph_ctrl.js]: What is the reason the endTime calculation needs to be moved here? Not that I think it's a bad idea to not always use the global $scope in all methods (adds so many side effects), more for my code understanding of the change of this PR.

File: app/assets/javascripts/angular/directives/graph_chart.js [hunk 3]
```diff
@@ -134,24 +276,24 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
...
-                scaleID = s.axisID;
-                break;
+                graphMax = axesBounds[matchingAxis.id].max;
+                bound.max = axesBounds[matchingAxis.id].max;
+              } else {
+                bound.max = maybeYMax;
+                graphMax = maybeYMax;
+              }
+              scaleID = s.axisID;
+              break;
```
Comment (grobie) [inline app/assets/javascripts/angular/directives/graph_chart.js]: All the line changes above look like broken indentation? Do you do these changes manual or is this your editor?

File: app/assets/javascripts/angular/directives/graph_chart.js [hunk 4]
```diff
@@ -359,9 +503,9 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
         if (scope.graphSettings.showLegend === "never" ||
             (scope.graphSettings.showLegend === "sometimes" && series.length > 5)) {
           $(element[0]).find(".legend").hide();
-          series.forEach(function(s) {
-            s.noLegend = true;
-          });
+        series.forEach(function(s) {
+          s.noLegend = true;
+        });
```
Comment (grobie) [inline app/assets/javascripts/angular/directives/graph_chart.js]: Again, broken indentation now?

File: app/assets/javascripts/angular/directives/graph_chart.js [hunk 5]
```diff
@@ -11,6 +12,147 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
...
+          rectangle.remove()
+        }
+
+        function onMousemove() {
+          //Draw the rectangle
+          var limits = [drag.startPX, d3.event.offsetX].sort();
+          rectangle
+          .attr("fill", "steelblue")
+          .attr("x", limits[0])
+          .attr("width", limits[1]-limits[0]);
```
Comment (grobie) [inline app/assets/javascripts/angular/directives/graph_chart.js]: Looks like a very unusual way to indent a multi-line statement.

File: app/assets/javascripts/angular/directives/graph_chart.js [hunk 6]
```diff
@@ -11,6 +12,147 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
...
+
+          drag.target = d3.event.target;
+          drag.startDt = pointAsDate(d3.event);
+          drag.y = d3.event.offsetY;
+          drag.startPX = d3.event.offsetX;
+          d3.event.preventDefault ? d3.event.preventDefault() : d3.event.returnValue = false;
+          d3.select(document).on("mousemove", onMousemove);
+          d3.select(document).on("mouseup", onMouseup);
+          d3.select(document).on("keyup", function() {
+            if (d3.event.keyCode === 27) {
```
Comment (grobie) [inline app/assets/javascripts/angular/directives/graph_chart.js]: Replace magic number with constant. http://www.refactoring.com/catalog/replaceMagicNumberWithSymbolicConstant.html

**Comment by grobie** [comment]
This is awesome!!!

I believe it's important to reload data after zooming in. If I zoom in into a specific time range, it's because I want to inspect the data in more detail. Maybe it's possible to do: zoom in -> show existing data immediately -> trigger data reload -> replace graph with more detailed data from server.

Another open question is whether zooming in in one graph should also change the time range in all other graphs. I'd argue it should, because if I'm interested in a specific graph, e.g. CPU usage, it's very likely I also want to zoom in the same time range in another dashboard graph, e.g. memory usage. This means zooming in would change the dashboard time settings. Open for discussion of course.

File: app/assets/javascripts/angular/controllers/graph_ctrl.js (hunk 2 above)
Comment (stuartnelson3) [inline app/assets/javascripts/angular/controllers/graph_ctrl.js]: Originally, I was going to call `refreshFn()` (on the next line) with updated arguments coming from the user's drag selection. Currently I'm thinking I will just update the `$scope.graph.endTime` and `$scope.graph.range` variables and call for a refresh, though, which makes this unnecessary.

In general I'm attempting to pass variables explicitly instead of defaulting to sending in `$scope` and then picking off what attributes I need, for clarity and to ease the job of extracting the graph drawing into its own independent library (something that is eventually going to happen, so we can consolidate graph functionality between prometheus and promdash).

File: app/assets/javascripts/angular/directives/graph_chart.js (hunk 5 above)
Comment (stuartnelson3) [inline app/assets/javascripts/angular/directives/graph_chart.js]: ah yeah, that's how my vim does it, and i was in "free-hack" mode, so i was just going with it

File: app/assets/javascripts/angular/directives/graph_chart.js (hunk 3 above)
Comment (stuartnelson3) [inline app/assets/javascripts/angular/directives/graph_chart.js]: ah sheet, yeah this is completely unrelated, my editor must have picked this up and messed it up.

**Comment by stuartnelson3** [comment]
Hmmmm, that seems like it could be a big assumption, especially given that there's (currently) no fast way to back out. Additionally, triggering a reload on all graphs could be slow. You're the main voice in the discussion, though, so it is mostly what solves your use-case.

**Comment by stuartnelson3** [comment]
getting some weird behavior when dragging across some "magic" boundary where the box stops rendering, will investigate more tomorrow.

File: app/assets/javascripts/angular/services/drag_selection.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    d3.select(document).on("mousemove", onMousemove);
+    d3.select(document).on("mouseup", onMouseup);
+    d3.select(document).on("keyup", function() {
+      if (d3.event.keyCode === ESCAPE_KEYCODE) {
+        reset();
+      }
+    });
+    d3.select(document).on("touchmove", onMousemove);
+    d3.select(document).on("touchend, touchcancel", onMouseup); // i think i can listen to both events like this..
+    // d3.select(document).on("touchcancel", onMouseup);
```
Comment (grobie) [inline app/assets/javascripts/angular/services/drag_selection.js]: Can you? If yes, cleanup comment?

Comment (stuartnelson3) [inline app/assets/javascripts/angular/services/drag_selection.js]: need to get my hands on an ipad at work, would love to test using touch for zooming :)

File: app/assets/javascripts/angular/controllers/graph_ctrl.js [hunk 8]
```diff
@@ -135,6 +138,9 @@ angular.module("Prometheus.controllers").controller('GraphCtrl',
   }
 
   $scope.refreshGraph();
+
+  $scope.drag = new DragSelection();
```
Comment (stuartnelson3) [inline app/assets/javascripts/angular/controllers/graph_ctrl.js]: this doesn't need to be created here; move it into the graph_chart and instantiate it there. also move the `setup()` function to the instantiation.

File: app/assets/javascripts/controllers.js [hunk 9]
```diff
@@ -17,6 +17,27 @@ Prometheus.Graph = {
     's': 1
   },
 
+  durationToString: function(seconds) {
+    // TODO(stuart): this is always returning seconds
+    // look to round to a certain amount based on the units
+    // in the graph? some level of precision lost.
```
Comment (stuartnelson3) [inline app/assets/javascripts/controllers.js]: This is up to debate. Currently, I have yet to see a range defined by dragging a selection on the graph return anything other than seconds. I'm all for removing this function and always returning seconds above (app/assets/javascripts/angular/resources/shared_graph_behavior.js:86). Up to you guys.

File: app/assets/javascripts/angular/directives/graph_chart.js [hunk 10]
```diff
@@ -277,6 +278,19 @@ angular.module("Prometheus.directives").directive('graphChart', ["$location", "W
         rsGraph.series.legend = legend;
         rsGraph.render();
 
+        var drag = new Rickshaw.Graph.DragZoom({
+          graph: rsGraph,
+          opacity: 0.5,
+          fill: 'steelblue',
+          minimumSelection: 25,
```
Comment (stuartnelson3) [inline app/assets/javascripts/angular/directives/graph_chart.js]: config is up to our preference. minimum selection is the number of pixels that a selection has to exceed in order for the highlighted area to be rendered. this protects against double clicking an area and having it render essentially nothing.

File: app/assets/javascripts/Rickshaw.Graph.DragZoom.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
+Rickshaw.namespace('Rickshaw.Graph.DragZoom');
```
Comment (stuartnelson3) [inline app/assets/javascripts/Rickshaw.Graph.DragZoom.js]: once this is all code reviewed and we're happy with how it works i'm planning on submitting it to the rickshaw repo.

**Comment by stuartnelson3** [comment]
@juliusv @grobie find some bugs for me

File: app/assets/javascripts/controllers.js (hunk 9 above)
Comment (grobie) [inline app/assets/javascripts/controllers.js]: Can you explain this a bit more? If I understand you correctly, it's possible that the input is given in years, days, etc. (month is not possible?)? 

Where does this input come from? Is this a rickshaw thing or related to promdash?

Comment (stuartnelson3) [inline app/assets/javascripts/controllers.js]: the input is given in seconds as a large integer. Then, that number is `%`'d against the seconds value to see if it is perfectly divisible by it. If so, then the value returned will be in those units. otherwise, it will be in seconds.
example:

`durationToString(259200)` => '3d'
259200 % 60*60*24 (this is `factors['d']`) === 0, so it is evenly divisible by days

it is used at app/assets/javascripts/angular/resources/shared_graph_behavior.js:86, seen in this PR.

The input comes from the rickshaw graph. The seconds value is the range in seconds highlighted by the user. Rickshaw manages X axis times behind the scenes in seconds.

**Comment by grobie** [comment]
:heart_eyes: This is so awesome. Super smooth experience, high five!

Now we need to talk about zooming out :D

One option would be to implement #289. This will essentially allow to maintain a history in the browser, so that `backspace` can be used to revert a step; therefore zooming out on backspace.

Another option would be to implement #147 adding for example a `z` shortcut to zoom out / revert an action.

**Comment by stuartnelson3** [comment]
Do you want to do that in this PR, or in a separate one?

I have an idea, but it exists externally to the code in this PR.

File: app/assets/javascripts/angular/directives/graph_chart.js (hunk 10 above)
Comment (grobie) [inline app/assets/javascripts/angular/directives/graph_chart.js]: Woud it be possible to use seconds as the minimum selection unit?

It's currently possible to zoom in into a total range <1s with values in ms as unit in the x axis. I guess it doesn't make sense to have a total range smaller than say 5s. Using time as absolute minimum instead of pixel width which is relative to the current state ensures zooming determines.

Comment (stuartnelson3) [inline app/assets/javascripts/angular/directives/graph_chart.js]: makes sense to me. I'll add the functionality and see what @juliusv thinks should be our minimum allowed zoom window

**Comment by stuartnelson3** [comment]
ping, opinion?

**Comment by juliusv** [comment]
Is there harm in allowing zoom windows all the way down to 1s?

**Comment by stuartnelson3** [comment]
my only concern is there wouldn't be valuable data at such a small time range

**Comment by juliusv** [comment]
Depends on the use case. Some people might run with 1s scraping. If it doesn't hurt, I wouldn't forbid it. They can always zoom out :)

**Comment by stuartnelson3** [comment]
zooming out via the "back" button may be fixed implicitly by #292

**Comment by grobie** [comment]
I'm vary about allowing such a small zoom range. It doesn't seem to make any sense with a default prometheus configuration (by default scrape interval is in the rang of 15s to 60s. Such a prometheus won't return any data points for a range of <5s). Though, I don't feel strong that we therefore have to forbid it.

:+1: if we're fine with back button / browser navigation for zooming out again. That can be discussed in the separate PR.

File: app/assets/javascripts/controllers.js [hunk 12]
```diff
@@ -17,6 +17,27 @@ Prometheus.Graph = {
     's': 1
   },
 
+  durationToString: function(seconds) {
+    // TODO(stuart): this is always returning seconds
+    // look to round to a certain amount based on the units
```
Comment (juliusv) [inline app/assets/javascripts/controllers.js]: Hmm, I can't parse this sentence somehow, or is a new one? (please use sentence-style caps and punctuation)

**Comment by juliusv** [comment]
This is magic and rocket science, I love it. Make it so! :+1: 

As for the minimum zoom, I don't feel strongly, so maybe 15s as a compromise is good.

File: app/assets/javascripts/controllers.js (hunk 12 above)
Comment (stuartnelson3) [inline app/assets/javascripts/controllers.js]: Ah sorry, I'll fix this. Basically, in broken no good engrish, it's saying "This function, for where it is being used, will always return a duration in seconds. Look at rounding to the nearest value." I don't think this is necessarily valid that we want to round, and it should be pretty self-evident that only perfectly divisible time units will be days/weeks/years, so I'm going to delete the comment.

**Comment by stuartnelson3** [comment]
I deferred to you regarding changing to 1 second, but my original reasoning follows what @grobie said: because of the standard scrape time, I really don't see the point in zooming in less than 15sec. I'll set it there, and make a note of it in the release to see if the customers have a strong opinion.

**Comment by stuartnelson3** [comment]
Also, regarding zoom-out: I think ctrl+z would be a good option, and I think that should be implemented whether or not selecting "back" in the browser works. But, as you said, discussion for another pr.

**Comment by stuartnelson3** [comment]
@juliusv do you have a preference on how to squash this? I'm thinking 2 commits: the first commit where I refactor the graphRefresher, then the rest of the commits (which is actually delightfully simple, as 99% of the code exists in a separate file).

File: app/assets/javascripts/Rickshaw.Graph.DragZoom.js (hunk 11 above)
Comment (stuartnelson3) [inline app/assets/javascripts/Rickshaw.Graph.DragZoom.js]: Note: I need to get some time with an iPad to get the touch swipe functionality working.

@juliusv are there any company ipads on the VPN (so they can access promdash)? it feels SO natural to look at a dashboard graph on an ipad, and swipe to zoom in. I would love to see if there's interest in adoption.

Comment (juliusv) [inline app/assets/javascripts/Rickshaw.Graph.DragZoom.js]: @stuartnelson3 I don't think so, but this would be an internal matter, best discussed maybe with mobile teams :) In general, I'd be surprised if there were many orgs out there that allowed sufficiently privileged VPN on iPads to use monitoring, but who knows...

**Comment by juliusv** [comment]
Yeah, I think that squashing strategy is fine.

**Comment by juliusv** [comment]
(and yes, zoom -> different PR)

File: app/assets/javascripts/Rickshaw.Graph.DragZoom.js (hunk 11 above)
Comment (grobie) [inline app/assets/javascripts/Rickshaw.Graph.DragZoom.js]: @stuartnelson3 What is the state on contributing this back to rickshaw? Still planned?

Comment (stuartnelson3) [inline app/assets/javascripts/Rickshaw.Graph.DragZoom.js]: still planned, but i haven't gotten around to it. will try to do soon.
